### ISSUE TITLE
Raspberry Pico W can't store link key 

### ISSUE THREAD

**Issue by wasdwasd0105** [bluekitchen/btstack#498]
**Describe the bug**
Hello,
I am writing a BTstack project with Pico W and want to save paired info. However, it seems BTstack may not utilize Pico's storage to save link key. gap_store_link_key_for_bd_addr was called to save link key. However, I get nothing when I use list_link_keys() to get the key

**To Reproduce**

Here is my project: https://github.com/wasdwasd0105/PicoW-usb2bt-audio

<img width="817" alt="Screenshot 2023-07-10 at 8 29 41 PM" src="https://github.com/bluekitchen/btstack/assets/12875499/3a146f01-d359-4d90-a160-ad28e91deca7">

**Expected behavior**

**Additional context**

**Comment by mringwal** [comment]
Hi. That's weird. BTstack uses it's btstack_tlv.h for persistent storage. On the Pico W, this is implemented in btstack_flash_bank.c, which provides pico_flash_bank_instance().

If you run the a2dp_sink_demo and connect from a mobile phone. Can you connect again after the pairing?

I would assume yes - I've played with the Cosmic Unicorn and BTstack a while ago and don't remember the need to pair every time.

@peterharperuk the list_link_keys() functions uses the link key iterator provided by our btstack_link_key_db_tlv.c, which only requires that btstack_tlv is properly set-up. The iteration uses the same logic as the regular look-up. (just to get you into the loop, no action item yet :)

**Comment by peterharperuk** [comment]
Yes, it should work. Pico W uses the last two blocks in flash to store the TLV stuff. It's possible to overwrite those, but it doesn't seem likely. Where your code that calls list_link_keys?

**Comment by mringwal** [comment]
@peterharperuk if you like, please copy list_link_keys from https://github.com/bluekitchen/btstack/blob/master/example/gap_link_keys.c#L64
into one of the classic examples and call it after pairing (or on start after a reboot..)

**Comment by wasdwasd0105** [comment]
<img width="958" alt="Screenshot 2023-07-14 at 11 03 34 PM" src="https://github.com/bluekitchen/btstack/assets/12875499/275f9865-a2d8-44e5-8907-91d727e389d7">

I think I got the reason: the storage system works well if I compile directly from https://github.com/raspberrypi/pico-examples/ and it will fail if I use a third-party temple such as https://github.com/MrGreensWorkshop/RasPiPicoSDK_BT_Classic_SPP. So what should I do to initiate the storage/DB part properly?

**Comment by wasdwasd0105** [comment]
And I found if I use a temple project, I can read the data but cant storage data

**Comment by mringwal** [comment]
Interesting find. We've only checked our examples on the Pico W. 
BTstack's TLV is initialized from cyw43_arch_init() which is also called from the main.c in MrGreensWorkshop. 

Could you start your app development from one of them? For SPP, please have a look at spp_counter or spp_streamer - the first one sends a counter, the second tries to send as fast as possible.

**Comment by sunnyqeen** [comment]
@wasdwasd0105
bool flash_safe_execute_core_init(void) is needed for multi-core
see  pico-sdk\src\rp2_common\pico_flash\include\pico\flash.h
I got similar problem.

**Comment by wasdwasd0105** [comment]
Yes! I solve the issue. I tried flash_safe_execute_core_init(void) on core 1 but still returned error -4. But I defined PICO_FLASH_ASSUME_CORE1_SAFE = 1 and it works. I use core 1 only to handle USB data so I don't use core 1 to write storage. Thanks!

**Comment by mringwal** [comment]
@peterharperuk Can you follow the problem and the solution/workaround? Is there a general fix that could be applied?

**Comment by wasdwasd0105** [comment]
I will leave the issue open
